File: frontend/js/views/commercial_lease.js
// Commercial (NNN) lease generator — base + triple-net charges, gross monthly
// rent, lease end date, via /calc/commercial-lease. Previews live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderCommercialLease(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.cl.h1.title">// COMMERCIAL LEASE (NNN)</span></h1>
        <p class="muted small" data-i18n="view.cl.hint.intro">
            Leases business space where the tenant pays base rent plus its share of operating costs
            (common-area maintenance, property tax, insurance). It computes the base and NNN charges from
            the per-square-foot rates and the area, the gross monthly rent, and the lease end date.
            Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.cl.h2.inputs">Lease details</h2>
            <form id="cl-form" class="inline-form">
                <label><span data-i18n="view.cl.label.state">State / jurisdiction</span>
                    <input type="text" name="state" value="Texas" required></label>
                <label><span data-i18n="view.cl.label.landlord">Landlord</span>
                    <input type="text" name="landlord_name" value=""></label>
                <label><span data-i18n="view.cl.label.tenant">Tenant (business)</span>
                    <input type="text" name="tenant_name" value=""></label>
                <label><span data-i18n="view.cl.label.premises">Premises address</span>
                    <input type="text" name="premises_address" value=""></label>
                <label><span data-i18n="view.cl.label.sqft">Square feet</span>
                    <input type="number" step="1" min="0" name="square_feet" value="2000" required></label>
                <label><span data-i18n="view.cl.label.base">Base rent ($/sq ft/yr)</span>
                    <input type="number" step="0.01" min="0" name="base_rent_psf_annual" value="30" required></label>
                <label><span data-i18n="view.cl.label.cam">CAM ($/sq ft/yr)</span>
                    <input type="number" step="0.01" min="0" name="cam_psf_annual" value="5"></label>
                <label><span data-i18n="view.cl.label.tax">Property tax ($/sq ft/yr)</span>
                    <input type="number" step="0.01" min="0" name="property_tax_psf_annual" value="3"></label>
                <label><span data-i18n="view.cl.label.ins">Insurance ($/sq ft/yr)</span>
                    <input type="number" step="0.01" min="0" name="insurance_psf_annual" value="2"></label>
                <label><span data-i18n="view.cl.label.start">Lease start date</span>
                    <input type="date" name="lease_start_date" value="2026-08-01" required></label>
                <label><span data-i18n="view.cl.label.term">Term (months)</span>
                    <input type="number" step="1" min="1" name="term_months" value="60" required></label>
                <label><span data-i18n="view.cl.label.use">Permitted use</span>
                    <input type="text" name="permitted_use" value="Retail store"></label>
                <label><span data-i18n="view.cl.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.cl.ph.statute'))}"></label>
            </form>
        </div>
        <div id="cl-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#cl-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            landlord_name: (fd.get('landlord_name') || '').trim(),
            tenant_name: (fd.get('tenant_name') || '').trim(),
            premises_address: (fd.get('premises_address') || '').trim(),
            square_feet: Number(fd.get('square_feet')) || 0,
            base_rent_psf_annual: Number(fd.get('base_rent_psf_annual')) || 0,
            cam_psf_annual: Number(fd.get('cam_psf_annual')) || 0,
            property_tax_psf_annual: Number(fd.get('property_tax_psf_annual')) || 0,
            insurance_psf_annual: Number(fd.get('insurance_psf_annual')) || 0,
            lease_start_date: fd.get('lease_start_date'),
            term_months: Math.round(Number(fd.get('term_months')) || 0),
            permitted_use: (fd.get('permitted_use') || '').trim(),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcCommercialLease(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.cl.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#cl-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.cl.card.gross">Gross rent / mo</div>
                    <div class="value">${money(doc.gross_monthly_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.cl.card.base">Base / mo</div>
                    <div class="value">${money(doc.base_monthly_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.cl.card.nnn">NNN / mo</div>
                    <div class="value">${money(doc.nnn_monthly_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.cl.card.end">Lease ends</div>
                    <div class="value">${esc(doc.lease_end_date || '—')}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="cl-copy" type="button" data-i18n="view.cl.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="cl-download" type="button" data-i18n="view.cl.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#cl-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.cl.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.cl.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#cl-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'commercial-lease.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/commercial_lease.rs
//! Commercial lease (triple-net / NNN) — leases business space where the tenant
//! pays base rent plus its share of the operating costs (common-area
//! maintenance, property tax, insurance). It computes the base and NNN charges
//! from the per-square-foot rates and the area, the gross monthly rent, and the
//! lease end date, then assembles the lease. Distinct from the residential
//! `lease` generator. Drafting aid, not legal advice.

use chrono::{Duration, Months, NaiveDate};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct CommercialLeaseInput {
    pub landlord_name: String,
    pub tenant_name: String,
    pub premises_address: String,
    pub square_feet: f64,
    /// Base rent, dollars per square foot per year.
    pub base_rent_psf_annual: f64,
    #[serde(default)]
    pub cam_psf_annual: f64,
    #[serde(default)]
    pub property_tax_psf_annual: f64,
    #[serde(default)]
    pub insurance_psf_annual: f64,
    pub lease_start_date: String,
    pub term_months: u32,
    /// Permitted use of the premises.
    #[serde(default)]
    pub permitted_use: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct CommercialLease {
    pub title: String,
    pub lease_start_date: String,
    pub lease_end_date: String,
    pub term_months: u32,
    pub base_annual_usd: f64,
    pub base_monthly_usd: f64,
    pub nnn_psf_annual: f64,
    pub nnn_annual_usd: f64,
    pub nnn_monthly_usd: f64,
    pub gross_annual_usd: f64,
    pub gross_monthly_usd: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &CommercialLeaseInput) -> CommercialLease {
    let base_annual = cents(i.square_feet * i.base_rent_psf_annual);
    let nnn_psf = i.cam_psf_annual + i.property_tax_psf_annual + i.insurance_psf_annual;
    let nnn_annual = cents(i.square_feet * nnn_psf);
    let gross_annual = cents(base_annual + nnn_annual);

    let base_monthly = cents(base_annual / 12.0);
    let nnn_monthly = cents(nnn_annual / 12.0);
    let gross_monthly = cents(gross_annual / 12.0);

    let end_date = NaiveDate::parse_from_str(&i.lease_start_date, "%Y-%m-%d")
        .ok()
        .and_then(|d| d.checked_add_months(Months::new(i.term_months)))
        .map(|d| (d - Duration::days(1)).format("%Y-%m-%d").to_string())
        .unwrap_or_default();

    let citation = i.statute_citation.trim();
    let pursuant = if citation.is_empty() {
        format!("This lease is governed by the laws of the State of {}.", i.state)
    } else {
        format!("This lease is governed by the laws of the State of {} ({}).", i.state, citation)
    };

    let nnn_body = if nnn_psf > 0.0 {
        format!(
            "In addition to base rent, the Tenant shall pay its triple-net share of operating costs at {} per square foot per year (CAM {}, property tax {}, insurance {}), totaling {} per year ({} per month).",
            money(nnn_psf),
            money(i.cam_psf_annual),
            money(i.property_tax_psf_annual),
            money(i.insurance_psf_annual),
            money(nnn_annual),
            money(nnn_monthly)
        )
    } else {
        "No separate triple-net charges apply; rent is gross.".to_string()
    };

    let use_body = if i.permitted_use.trim().is_empty() {
        "The premises shall be used only for lawful commercial purposes.".to_string()
    } else {
        format!("The premises shall be used only for: {}.", i.permitted_use.trim())
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!("Landlord: {}\nTenant: {}\nPremises: {} ({} sq ft)", i.landlord_name, i.tenant_name, i.premises_address, i.square_feet),
        },
        DocClause {
            heading: "1. Term".into(),
            body: format!("The lease term is {} months, beginning {} and ending {}.", i.term_months, i.lease_start_date, end_date),
        },
        DocClause {
            heading: "2. Base Rent".into(),
            body: format!(
                "Base rent is {} per square foot per year on {} square feet, or {} per year ({} per month).",
                money(i.base_rent_psf_annual), i.square_feet, money(base_annual), money(base_monthly)
            ),
        },
        DocClause { heading: "3. Triple-Net Charges".into(), body: nnn_body },
        DocClause {
            heading: "4. Total Rent".into(),
            body: format!(
                "Total rent is {} per year, payable in monthly installments of {}.",
                money(gross_annual), money(gross_monthly)
            ),
        },
        DocClause { heading: "5. Use".into(), body: use_body },
        DocClause { heading: "6. Governing Law".into(), body: pursuant },
        DocClause {
            heading: "Signatures".into(),
            body: format!(
                "Landlord: ____________________  Date: __________\n{}\n\nTenant: ____________________  Date: __________\n{}",
                i.landlord_name, i.tenant_name
            ),
        },
    ];

    CommercialLease {
        title: "Commercial Lease Agreement (NNN)".into(),
        lease_start_date: i.lease_start_date.clone(),
        lease_end_date: end_date,
        term_months: i.term_months,
        base_annual_usd: base_annual,
        base_monthly_usd: base_monthly,
        nnn_psf_annual: nnn_psf,
        nnn_annual_usd: nnn_annual,
        nnn_monthly_usd: nnn_monthly,
        gross_annual_usd: gross_annual,
        gross_monthly_usd: gross_monthly,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> CommercialLeaseInput {
        CommercialLeaseInput {
            landlord_name: "Prop Holdings".into(),
            tenant_name: "Beta Corp".into(),
            premises_address: "100 Commerce Plaza, Suite 5".into(),
            square_feet: 2000.0,
            base_rent_psf_annual: 30.0,
            cam_psf_annual: 5.0,
            property_tax_psf_annual: 3.0,
            insurance_psf_annual: 2.0,
            lease_start_date: "2026-08-01".into(),
            term_months: 60,
            permitted_use: "Retail store".into(),
            state: "Texas".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn base_rent_annual_and_monthly() {
        let d = generate(&base());
        assert!(close(d.base_annual_usd, 60_000.0));
        assert!(close(d.base_monthly_usd, 5_000.0));
    }

    #[test]
    fn nnn_charges() {
        let d = generate(&base());
        assert!(close(d.nnn_psf_annual, 10.0));
        assert!(close(d.nnn_annual_usd, 20_000.0));
        assert!(close(d.nnn_monthly_usd, 1_666.67));
    }

    #[test]
    fn gross_rent() {
        let d = generate(&base());
        assert!(close(d.gross_annual_usd, 80_000.0));
        assert!(close(d.gross_monthly_usd, 6_666.67));
    }

    #[test]
    fn end_date_is_start_plus_term_minus_day() {
        // 2026-08-01 + 60 months − 1 day = 2031-07-31.
        assert_eq!(generate(&base()).lease_end_date, "2031-07-31");
    }

    #[test]
    fn no_nnn_is_gross_lease() {
        let d = generate(&CommercialLeaseInput {
            cam_psf_annual: 0.0,
            property_tax_psf_annual: 0.0,
            insurance_psf_annual: 0.0,
            ..base()
        });
        assert!(close(d.nnn_annual_usd, 0.0));
        assert!(close(d.gross_annual_usd, 60_000.0));
        let c = d.clauses.iter().find(|c| c.heading.contains("Triple-Net")).unwrap();
        assert!(c.body.contains("rent is gross"));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&CommercialLeaseInput { statute_citation: "Tex. Prop. Code".into(), ..base() });
        assert_eq!(d.statutory_citation, "Tex. Prop. Code");
        assert!(d.clauses.iter().any(|c| c.body.contains("Tex. Prop. Code")));
    }
}
